ambry-store/src/test/java/com.github.ambry.store/CompactionPolicyTest.java

```diff
@@ -0,0 +1,365 @@
+/**
+ * Copyright 2017 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
+
+import com.codahale.metrics.MetricRegistry;
+import com.github.ambry.config.StoreConfig;
+import com.github.ambry.config.VerifiableProperties;
+import com.github.ambry.utils.MockTime;
+import com.github.ambry.utils.Pair;
+import com.github.ambry.utils.TestUtils;
+import com.github.ambry.utils.Time;
+import com.github.ambry.utils.UtilsTest;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.NavigableMap;
+import java.util.Properties;
+import java.util.TreeMap;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import static org.junit.Assert.*;
+
+
+/**
+ * Unit tests different {@link CompactionPolicy}
+ */
+@RunWith(Parameterized.class)
+public class CompactionPolicyTest {
+
+  static long CAPACITY_IN_BYTES = 10 * 1024 * 1024;
+  private static long SEGMENT_CAPACITY_IN_BYTES = CAPACITY_IN_BYTES / 10;
+  private static long DEFAULT_USED_CAPACITY_IN_BYTES = CAPACITY_IN_BYTES * 6 / 10;
+  private static long DEFAULT_MAX_BLOB_SIZE = CAPACITY_IN_BYTES / 100;
+  private static long SEGMENT_HEADER_SIZE = CAPACITY_IN_BYTES / 50;
+  // the properties that will used to generate a StoreConfig. Clear before use if required.
+  private final Properties properties = new Properties();
+  private Time time;
+  private MockBlobStore blobStore;
+  private StoreConfig config;
+  private long messageRetentionTimeInMs;
+  private MockBlobStoreStats mockBlobStoreStats;
+  private final CompactionPolicyType type;
+  private CompactionPolicy compactionPolicy;
+
+  /**
+   * Refers to different types of {@link CompactionPolicy}
+   */
+  enum CompactionPolicyType {
+    DEFAULT,
```

Comment:
why not use the factories instead? The parameters can be the class name of the factories (just like in config).